lib/redirect-agent.js

```diff
@@ -0,0 +1,169 @@
+const { InvalidArgumentError } = require('./core/errors')
+const util = require('./core/util')
+const EventEmitter = require('events')
+const { getGlobalAgent } = require('./agent')
+const assert = require('assert')
+
+const kMaxRedirections = Symbol('maxRedirections')
+const kAgent = require('agent')
+
+module.exports = class RedirectAgent extends EventEmitter {
+  constructor ({ maxRedirections = 10, agent = getGlobalAgent() } = {}) {
+    super()
+
+    if (maxRedirections != null && (maxRedirections < 0 || !Number.isInteger(maxRedirections))) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    this[kMaxRedirections] = maxRedirections
+    this[kAgent] = agent
+
+    this[kAgent].on('connect', (...args) => this.emit(...args))
+    this[kAgent].on('disconnect', (...args) => this.emit(...args))
+  }
+
+  dispatch (origin, { maxRedirections = this[kMaxRedirections], ...opts }, handler) {
+    if (maxRedirections != null && (maxRedirections < 0 || !Number.isInteger(maxRedirections))) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    if (!maxRedirections) {
+      return this[kAgent].dispatch(origin, opts, handler)
+    }
+
+    if (util.isStream(opts.body) && util.bodyLength(opts.body) === 0) {
+      opts.body
+        .on('data', /* istanbul ignore next */ function () {
+          /* istanbul ignore next */
+          assert(false)
+        })
+        .on('error', function (err) {
+          // TODO (fix): What to do?
+          throw err
+        })
+        .on('end', function () {
+          util.destroy(this)
+        })
+      opts.body = null
+    }
+
+    if (util.isStream(opts.body)) {
+      return this[kAgent].dispatch(origin, opts, handler)
+    }
+
+    opts.maxRedirections = maxRedirections - 1
+
+    this[kAgent].dispatch(origin, opts, {
+      agent: this,
+      location: null,
+      opts,
+      origin,
+      handler,
+
+      onConnect (abort) {
+        this.onConnect(abort)
+      },
+
+      onUpgrade (statusCode, headers, socket) {
+        this.onUpgrade(statusCode, headers, socket)
+      },
+
+      onError (error) {
+        this.onError(error)
+      },
+
+      onHeaders (statusCode, headers, resume) {
+        if ([300, 301, 302, 303, 307, 308].indexOf(statusCode) === -1) {
+          return this.handler.onHeaders(statusCode, headers, resume)
+        }
+
+        this.location = parseLocation(headers)
+
+        if (!this.location) {
+          return this.handler.onHeaders(statusCode, headers, resume)
+        }
+
+        this.location = new URL(this.location, this.origin)
+
+        if (!this.opts.history) {
+          opts.history = new Set()
```

Comment:
I wouldn't do this. This will make impossible for users to detect if there was any redirect cycle.
Let the users deal with it. Since we have maxRedirections we don't really care.

Reply:
How will the user deal with it though? We would need some kind of hook in that case.

Reply:
Well, we have `maxRedirections`, isn't? If they get in a cycle at some point it will stop following ending with a 3xx.
And then they'll check history in the callback data to deal with it. But at undici level we don't care. Or not?

Reply:
That's a valid point. Where do we put the history though. There is no generic callback here. Does the http standard provide any header (such as `"via"`) where we can put the history?

@mcollina wdyt?

Reply:
As you can see in my PR, I expanded `onHeaders` in all resources so that can accept additional arguments. And I used to pass the history as it builds the callback data for all methods.

This probably is also a good idea for future expandability as new `Agent`s (either in undici, or on user-level) might benefit for it to attach custom callback data.

Reply:
We agreed we would do this in a follow up PR.